File: PathoFroh/src/main/java/com/patho/main/repository/service/TaskRepositoryCustom.java
package com.patho.main.repository.service;

import java.util.Calendar;
import java.util.List;
import java.util.Optional;

import javax.persistence.criteria.CriteriaQuery;
import javax.persistence.criteria.Predicate;
import javax.persistence.criteria.Root;

import com.patho.main.model.patient.Task;
import com.patho.main.util.worklist.search.WorklistSearchExtended;

public interface TaskRepositoryCustom {

    public Optional<Task> findOptionalByIdAndInitialize(Task task, boolean loadCouncils, boolean loadDiangoses, boolean loadPDFs,
                                                        boolean loadContacts, boolean loadParent);

    public Optional<Task> findOptionalByIdAndInitialize(Long id, boolean loadCouncils, boolean loadDiangoses, boolean loadPDFs,
                                                        boolean loadContacts, boolean loadParent);

    public Optional<Task> findOptionalByTaskId(String taskID);

    public Optional<Task> findOptionalByTaskId(String taskID, boolean loadCouncils, boolean loadDiangoses, boolean loadPDFs,
                                               boolean loadContacts, boolean loadParent);

    public Optional<Task> findOptionalBySlideID(String taskID, int uniqueSlideIDInBlock, boolean loadCouncils, boolean loadDiangoses,
                                                boolean loadPDFs, boolean loadContacts, boolean loadParent);

    public Optional<Task> findOptinalByLastID(Calendar ofYear, boolean loadCouncils, boolean loadDiangoses, boolean loadPDFs,
                                              boolean loadContacts, boolean loadParent);

    public List<Task> findByCriteria(WorklistSearchExtended worklistSearchExtended, boolean initParent);

    public Optional<Task> find(CriteriaQuery<Task> criteria, Root<Task> root, List<Predicate> predicates, boolean loadCouncils,
                               boolean loadDiangoses, boolean loadPDFs, boolean loadContacts, boolean loadParent);

}

File: PathoFroh/src/main/kotlin/com/patho/main/repository/jpa/custom/TaskRepositoryCustom.kt
package com.patho.main.repository.jpa.custom

import com.patho.main.model.patient.Task
import com.patho.main.util.exceptions.TaskNotFoundException
import com.patho.main.util.search.settings.ExtendedSearch
import java.time.LocalDate
import javax.persistence.criteria.CriteriaQuery
import javax.persistence.criteria.Predicate
import javax.persistence.criteria.Root

interface TaskRepositoryCustom {

    fun findByID(task: Task, loadCouncils: Boolean, loadDiagnoses: Boolean,
                 loadPDFs: Boolean, loadContacts: Boolean, loadParent: Boolean): Task

    fun findByID(id: Long, loadCouncils: Boolean, loadDiagnoses: Boolean,
                 loadPDFs: Boolean, loadContacts: Boolean, loadParent: Boolean): Task

    @Throws(TaskNotFoundException::class)
    fun findByTaskID(taskID: String, loadCouncils: Boolean = false, loadDiagnoses: Boolean = false,
                     loadPDFs: Boolean = false, loadContacts: Boolean = false, loadParent: Boolean = false): Task

    @Throws(TaskNotFoundException::class)
    fun findByLastID(ofYear: LocalDate, loadCouncils: Boolean, loadDiagnoses: Boolean, loadPDFs: Boolean,
                     loadContacts: Boolean, loadParent: Boolean): Task

    @Throws(TaskNotFoundException::class)
    fun findBySlideID(taskID: String, uniqueSlideIDInBlock: Int, loadCouncils: Boolean,
                      loadDiagnoses: Boolean, loadPDFs: Boolean, loadContacts: Boolean, loadParent: Boolean): Task

    fun findByExtendedSearchSettings(extendedSearch: ExtendedSearch, loadCouncils: Boolean,
                                     loadDiagnoses: Boolean, loadPDFs: Boolean, loadContacts: Boolean, loadParent: Boolean): List<Task>

    fun findAllByID(ids: List<Long>, loadCouncils: Boolean,
                    loadDiagnoses: Boolean, loadPDFs: Boolean, loadContacts: Boolean, loadParent: Boolean): List<Task>

    fun findAll(criteria: CriteriaQuery<Task>, root: Root<Task>, predicates: MutableList<Predicate>,
                loadCouncils: Boolean, loadDiagnoses: Boolean, loadPDFs: Boolean, loadContacts: Boolean, loadParent: Boolean): List<Task>
}